ARROW-16407: [R] Extend `parse_date_time` to cover hour, dates, and minutes components

dragosmg opened the pull request:
This PR improves `parse_date_time()` by:
* adding support for orders with the hours, minutes, and seconds components
* adding support for unseparated strings ([ARROW-16446](https://issues.apache.org/jira/browse/ARROW-16446))
* supporting the `exact` argument:
    * allows users to pass `exact = TRUE` in which case the `orders` are taken as they are (they are considered `formats` and passed to `strptime`)
    * `exact = FALSE` implies `formats` are derived from `orders`
* allowing the `truncated` argument
    * denotes number of formats that might be missing. For example, passing an `order` like `ymd_HMS` and a value of 1 for `truncated` will attempt parsing with both `ymd_HMS` and `ymd_HM` orders  
* erroring when the user passes `quiet = FALSE` 
* improves the utility function used to generate `formats` (which are then passed on to `strptime`) from `orders`
    * less hard-coding and increased ability to deal with different orders and separators 

the `ymd HMS` orders (and variants) will parse correctly:
``` r
library(dplyr, warn.conflicts = FALSE)
library(lubridate, warn.conflicts = FALSE)
library(arrow, warn.conflicts = FALSE)

test_df <- tibble(
  x = c("2011-12-31 12:59:59", "2010-01-01 12:11", "2010-01-01 12", "2010-01-01")
)

test_df %>%
  mutate(
    y = parse_date_time(x, "Ymd HMS", truncated = 3)
  )
#> # A tibble: 4 × 2
#>   x                   y                  
#>   <chr>               <dttm>             
#> 1 2011-12-31 12:59:59 2011-12-31 12:59:59
#> 2 2010-01-01 12:11    2010-01-01 12:11:00
#> 3 2010-01-01 12       2010-01-01 12:00:00
#> 4 2010-01-01          2010-01-01 00:00:00

test_df %>%
  arrow_table() %>% 
  mutate(
    y = parse_date_time(x, "Ymd HMS", truncated = 3)
  ) %>% 
  collect()
#> # A tibble: 4 × 2
#>   x                   y                  
#>   <chr>               <dttm>             
#> 1 2011-12-31 12:59:59 2011-12-31 12:59:59
#> 2 2010-01-01 12:11    2010-01-01 12:11:00
#> 3 2010-01-01 12       2010-01-01 12:00:00
#> 4 2010-01-01          2010-01-01 00:00:00
```

<sup>Created on 2022-05-19 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

`exact = TRUE` can also be used:
<details>

``` r
library(arrow, warn.conflicts = FALSE)
library(lubridate, warn.conflicts = FALSE)
library(dplyr, warn.conflicts = FALSE)

test_df <- tibble(
  x = c("11/23/1998 07:00:00", "6/18/1952 0135", "2/25/1974 0523", "9/07/1985 01", NA)
)

test_df %>%
  mutate(
    parsed_x =
      parse_date_time(
        x,
        c("%m/%d/%Y %I:%M:%S", "%m/%d/%Y %H%M", "%m/%d/%Y %H"),
        exact = TRUE
      )
  )
#> # A tibble: 5 × 2
#>   x                   parsed_x           
#>   <chr>               <dttm>             
#> 1 11/23/1998 07:00:00 1998-11-23 07:00:00
#> 2 6/18/1952 0135      1952-06-18 01:35:00
#> 3 2/25/1974 0523      1974-02-25 05:23:00
#> 4 9/07/1985 01        1985-09-07 01:00:00
#> 5 <NA>                NA

test_df %>%
  arrow_table() %>% 
  mutate(
    parsed_x =
      parse_date_time(
        x,
        c("%m/%d/%Y %I:%M:%S", "%m/%d/%Y %H%M", "%m/%d/%Y %H"),
        exact = TRUE
      )
  ) %>%
  collect()
#> # A tibble: 5 × 2
#>   x                   parsed_x           
#>   <chr>               <dttm>             
#> 1 11/23/1998 07:00:00 1998-11-23 07:00:00
#> 2 6/18/1952 0135      1952-06-18 01:35:00
#> 3 2/25/1974 0523      1974-02-25 05:23:00
#> 4 9/07/1985 01        1985-09-07 01:00:00
#> 5 <NA>                NA
```

<sup>Created on 2022-05-20 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>
</details>

jonkeane commented on r/R/dplyr-datetime-helpers.R:
```diff
@@ -201,19 +213,130 @@ build_formats <- function(orders) {
+    "y" = c("%y", "%Y"),
+    "m" = c("%m", "%B", "%b"),
+    "d" = "%d",
+    "H" = "%H",
```
Do we also want | need `%I` here?

dragosmg replied:
Done

jonkeane commented on r/R/dplyr-datetime-helpers.R:
```diff
@@ -201,19 +213,130 @@ build_formats <- function(orders) {
+  split_order <- strsplit(order, split = "")[[1]]
+
+  outcome <- expand.grid(char_list[split_order])
+  formats_with_sep <- do.call(paste, c(outcome, sep = "-"))
+  formats_without_sep <- do.call(paste, c(outcome, sep = ""))
+  c(formats_with_sep, formats_without_sep)
```
This would be something we would want to measure with a benchmark on a bunch of data, but something to consider is instead of doing both with and without separators on all, you could do something like `ifelse(grepl("-", col), formats_with_sep, formats_without_sep)` — though not quite right here, since you would want that to be passed as an expression

dragosmg replied:
What is the data has a mixture of both? I imagine it could happen in very messy datasets. So probably more like:
```r
ifelse(grepl("-", col), try_both_formats, try_only_without_sep) 
```

jonkeane replied:
`ifelse(grepl(...), ...)` should be operating at the element level for the column (e.g. it should be vectorized) so that there might be a mixture shouldn't be a problem.

dragosmg replied:
Using `ifelse` conditional on `x` containing a `"-"` returns an Expression. Currently I am using `purrr::map()` to build a list of Expressions, which would not work in this instance.

This is a sketch of the logic showing where the `ifelse` approach would trip up.
```r
formats <- call_binding(
  "if_else",
  call_binding("grepl", "-", x),
  formats_with_sep,
  formats_without_sep
  )

# now formats is an Expression and we cannot map over it; this would fail
  map(
    formats,
    ~ build_expr(
      "strptime",
      x,
      options = list(format = .x, unit = 0L, error_is_null = TRUE)
    )
  )
```

dragosmg replied:
In conclusion, I propose `formats` stays as-is until we have a way to iterate over Expressions.

jonkeane replied:
Right, you would need to add that `if_else()` to the expressions. This will require this code be in a slightly different location to take be added in the right spot (though you should be able to construct a test to see what kinds of performance impacts there are for running 1x orders versus 2x in this case).

dragosmg replied:
I managed to apply the essence of this, but in a slightly different way: 
* return a single version of `formats` - with separators
* map over them, but remote the separator if `x` doesn't have one

dragosmg replied:
@jonkeane I'm going to mark this as **done**. I haven't run any tests, but the approach makes a lot of sense. Happy to have a follow-up ticket and see if this move actually had a positive impact on performance. I don't expect it to have a negative impact though.

jonkeane commented on r/R/dplyr-datetime-helpers.R:
```diff
@@ -201,19 +213,130 @@ build_formats <- function(orders) {
+    )
+  }
+
+  # for `yq` we need to transform the quarter into the start month (lubridate
+  # behaviour) and then add 01 to parse to the first day of the quarter
+  augmented_x_yq <- NULL
+  if (any(orders == "yq")) {
+    # extract everything that comes after the `-` separator, i.e. the quarter
+    # (e.g. 4 from 2022-4)
+    quarter_x <- call_binding("gsub", "^.*?-", "", processed_x)
+    # we should probably error if quarter is not in 1:4
+    # extract everything that comes before the `-`, i.e. the year (e.g. 2002
+    # in 2002-4)
+    year_x <- call_binding("gsub", "-.*$", "", processed_x)
+    quarter_x <- quarter_x$cast(int32())
+    month_x <- (quarter_x - 1) * 3 + 1
+    augmented_x_yq <- call_binding("paste0", year_x, "-", month_x, "-01")
+  }
```
Do we have a jira for parsing something like `2020-01` into `2020-01-01` in C++ already? I'm not 100% sure we want to push that kind of logic into C++, but it would be good to if other languages do it as well. If we do, we should list it here to remind ourselves we can get rid of this hack when that issue is done

dragosmg replied:
I wasn't able to find one. This behaviour is surprising however:

``` r
library(arrow, warn.conflicts = FALSE)

call_function(
  "strptime",
  Array$create("2020-01"),
  options = list(
    format = "%Y-%m",
    unit = 0L
  )
)
#> Array
#> <timestamp[s]>
#> [
#>   2019-12-31 00:00:00
#> ]
```

<sup>Created on 2022-05-23 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

dragosmg replied:
Done. https://issues.apache.org/jira/browse/ARROW-16627

jonkeane commented on r/R/dplyr-datetime-helpers.R:
```diff
@@ -201,19 +213,130 @@ build_formats <- function(orders) {
+    # this is the situation in which orders were passed with `exact = TRUE`
+    # no data processing takes place
+    # we don't derive formats as the orders are assumed to be formats
+    parse_attempt_expressions <- build_strptime_exps(x, orders)
+    return(parse_attempt_expressions)
```
```suggestion
    return(build_strptime_exps(x, orders))
```

Slightly more simplified?

dragosmg replied:
I think we can simplify this even further, by calling `build_strptime_exprs` inside `parse_date_time` on the `exact == TRUE` branch:
```r
    if (exact == TRUE) {
      parse_attempts <- build_strptime_exps(x, orders)
    } 
```

https://github.com/dragosmg/arrow/blob/628b4f83bbb75fd7cf5c79862cb1000ba95bca7a/r/R/dplyr-funcs-datetime.R#L504-L505

dragosmg replied:
Done

jonkeane commented on r/R/dplyr-funcs-datetime.R:
```diff
@@ -488,105 +488,28 @@ register_bindings_duration_helpers <- function() {
-      )
+    if (truncated != 0) {
+      # build several orders for truncated formats
+      orders <- map_chr(0:truncated, ~ substr(orders, start = 1, stop = nchar(orders) - .x))
```
Do we have a test where `truncated` is passed and is larger than the smallest order? Something like `lubridate::parse_date_time("2020-01", orders = "Ymd", truncate = 10)`

dragosmg replied:
Nope. I was thinking to maybe limit `truncated` to 3 (i.e. error if greater). What do you think?

jonkeane replied:
We shouldn't rely on a hard coded limit — even with 3, we might still run into the same issue I'm alluding to here: when the number provided is larger than the entire string of formats.

dragosmg replied:
I was thinking 3 was a good number, but in relative terms `nchar(order)` as the upper limit would be too high. We need a few characters left (at least 3) to build a `format`. So `nchar(order) - max(truncated) >= 3`, which gives a limit for truncated of `nchar(orders) - 3`.

dragosmg replied:
Done.

jonkeane commented on r/tests/testthat/test-dplyr-funcs-datetime.R:
```diff
@@ -1942,3 +1925,206 @@ test_that("lubridate's fast_strptime", {
+        ) %>%
+        collect(),
+      "`quiet = FALSE` not supported in Arrow"
+    )
+  )
```
Should we assert the second warning here?

dragosmg replied:
Done

jonkeane commented on r/R/dplyr-datetime-helpers.R:
```diff
@@ -201,19 +213,130 @@ build_formats <- function(orders) {
+  processed_x <- processed_data[["processed_x"]]
+  augmented_x_ym <- processed_data[["augmented_x_ym"]]
+  augmented_x_yq <- processed_data[["augmented_x_yq"]]
+
+  # build a list of expressions for each format
+  parse_attempt_exp_processed_x <- build_strptime_exps(processed_x, formats)
+
+  # build separate expression lists of parsing attempts for the orders that
+  # need an augmented `x`
+  # list for attempts when orders %in% c("ym", "my")
+  parse_attempt_exp_augmented_x_ym <- list()
+
+  if (!is.null(augmented_x_ym)) {
+    parse_attempt_exp_augmented_x_ym <- build_strptime_exps(augmented_x_ym, formats)
+  }
+
+  # list for attempts when orders %in% c("yq")
+  parse_attempt_exp_augmented_x_yq <- list()
+  if (!is.null(augmented_x_yq)) {
+    parse_attempt_exp_augmented_x_yq <- build_strptime_exps(augmented_x_yq, formats)
+  }
+
+  # combine all attempts expressions in prep for coalesce
+  # if the users passes only a short order (`ym`, `my` or `yq`) then only use
+  # the corresponding augmented_x
+  if (all(orders == "ym") || all(orders == "my")) {
+    parse_attempt_expressions <- parse_attempt_exp_augmented_x_ym
+  } else if (all(orders == "yq")) {
+    parse_attempt_expressions <- parse_attempt_exp_augmented_x_yq
+  } else {
+    parse_attempt_expressions <- c(
+      # if we have an augmented x give preference to the corresponding
+      # parsing attempts
+      parse_attempt_exp_augmented_x_ym,
+      parse_attempt_exp_augmented_x_yq,
+      parse_attempt_exp_processed_x
+    )
+  }
+  parse_attempt_expressions
+}
```
This is a pretty hairy chunk that's pretty hard to follow. We should make it a bit easier on our future selves to follow and maintain it by doing some combination of (or all) of the following:

* Adding slightly more detailed comments, especially explaining what general shape of things might be in the arguments — it's hard to know now what things are lists of expressions versus lists of characters like `"Ym"` versus  lists of characters like `"%Y-%m"`
* Refactor this code to work on more directly, instead of having sets of functions that produce lists that you need to subset out of, would it be possible to have functions that take as input (vectors of) single elements and operate on them to go from `"Ym"` to the expression, expanding as necessary.

dragosmg replied:
💯 You should have seen `parse_date_time` before the refactoring. I'll try to refactor even more.

dragosmg replied:
Done.

jonkeane reviewed (changes requested):
A few comments for you

jonkeane commented on r/R/dplyr-datetime-helpers.R:
```diff
@@ -201,19 +218,100 @@ build_formats <- function(orders) {
+attempt_parsing <- function(x,
+                            orders) {
+  # translate orders into possible formats
+  formats <- build_formats(orders)
+
+  processed_data <- process_data_for_parsing(x, orders)
+
+  parse_attempt_exprs_list <- map(processed_data, build_strptime_exprs, formats)
```
Could you add some more comments describing what are in each of these variables?

dragosmg replied:
Done

jonkeane commented on r/R/dplyr-datetime-helpers.R:
```diff
@@ -201,19 +218,100 @@ build_formats <- function(orders) {
+    parse_attempt_exprs_list$processed_x <- list()
+  }
+
+  purrr::flatten(parse_attempt_exprs_list)
```
What's the structure here that needs flattening? I'm not saying it's wrong — I'm just not following from the code above why this is necesary.

jonkeane commented on r/R/dplyr-datetime-helpers.R:
```diff
@@ -179,7 +179,24 @@ build_formats <- function(orders) {
+    "mdy_HM", "mdy_H", "ydm_HMS", "ydm_HM", "ydm_H"
+  )
+  # support "%I" hour formats
+  ymd_ims_orders <- gsub("H", "I", ymd_hms_orders)
```
Is this ok to do? I thought `H` was 24-hour hours and `I` was 12? Do we have a test for using H with 24-hour style hours?

dragosmg replied:
Yes, this is used to add to the whitelist of supported orders, based off an already hardcoded list containing `"H"`. See my other [comment](https://github.com/apache/arrow/pull/13196#discussion_r880193852). We could always remove this code and figure out a different way to message the user. 
The whitelist approach was introduced as response to [this bit of feedback](https://github.com/apache/arrow/pull/12589#discussion_r861780324).

dragosmg replied:
You mean, tests for using `"I"` with 24h?

dragosmg replied:
Added a [test](https://github.com/dragosmg/arrow/blob/4846f806328c3f63f9d3aaca24ad14f22aec571f/r/tests/testthat/test-dplyr-funcs-datetime.R#L2027) for the situation when `"I"` is greater than 12.

jonkeane reviewed (changes requested):
This is moving, still a few functionality rough edges that need work. A few general notes:

* I'm still struggling a bit following some of the logic. More comments, especially with details about what's in some of the various variables would be helpful to follow what's going on. It's hard to know now what things are lists of expressions versus lists of characters like "Ym" versus lists of characters like "%Y-%m"

* Should we test `build_formats()` directly to detect some of the more bespoke | full set of order checks? This would make the tests a lot easier to reason about (what we're testing is more obvious + we wouldn't need to have a whole data frame + date setup in it) Take a look at the other helper functions to see if the same is true for them.

dragosmg commented on r/R/dplyr-datetime-helpers.R:
```diff
@@ -179,7 +179,24 @@ build_formats <- function(orders) {
+  supported_orders <- c(
+    ymd_orders,
+    ymd_hms_orders,
+    gsub("_", " ", ymd_hms_orders), # allow "_", " " and "" as separators
```
I built the parsing functionality with a distinction between `orders` and `formats`. `orders` are what the users supply (can be both with and without `"%"`) and `formats` are those we derive from the supplied `orders`. `supported_orders` is pretty much a whitelist of orders we can parse. I defined it to make it easier to return a meaningful message to the user when they supply an order we do not yet support. You are right in terms of `formats`, but the snippet of code deals with `orders`, which can be have separators that are not found in the data - e.g. the `ymd_HMS` order should parse correctly `"2022-05-23 12:34:56"` even though the string does not have `"_"` in it.

dragosmg replied:
We could remove the `orders` whitelist and try to come up with a different way of surfacing meaningful error messages.

dragosmg replied:
See [my comment](https://github.com/apache/arrow/pull/13196#issuecomment-1137032565): I think we should have a follow-up issue in which we tie up any loose ends.

dragosmg commented on r/R/dplyr-datetime-helpers.R:
```diff
@@ -201,19 +218,100 @@ build_formats <- function(orders) {
+  # for `yq` we need to transform the quarter into the start month (lubridate
+  # behaviour) and then add 01 to parse to the first day of the quarter
+  augmented_x_yq <- NULL
+  if (any(orders == "yq")) {
+    # extract everything that comes after the `-` separator, i.e. the quarter
+    # (e.g. 4 from 2022-4)
+    quarter_x <- call_binding("gsub", "^.*?-", "", processed_x)
+    # we should probably error if quarter is not in 1:4
+    # extract everything that comes before the `-`, i.e. the year (e.g. 2002
+    # in 2002-4)
+    year_x <- call_binding("gsub", "-.*$", "", processed_x)
+    quarter_x <- quarter_x$cast(int32())
+    month_x <- (quarter_x - 1) * 3 + 1
+    augmented_x_yq <- call_binding("paste0", year_x, "-", month_x, "-01")
+  }
```
Done. I've added support and unit tests for the `"qy"` order.

dragosmg commented:
Given this is a relatively large PR I propose I open a follow-up ticket to go with a fine-tooth comb and check for edge cases, formats , etc. I think we cover the vast majority of use cases, but I'd like to be able to spend some time looking  at unit tests, helpers descriptions, error messages etc.

dragosmg commented:
I think this is ready for another look @jonkeane. What changed since our last conversation?
* I benchmarked my original implementation (formats with and without separators in a single vector) against versions in which the formats were separated (see [this comment](https://github.com/apache/arrow/pull/13196#issuecomment-1161550157) for more details). The initial implementation proved faster and so I went with that approach.
* I added roxygen2 headers for the helper functions to give more clarity what is happening where
* I added a suite of unit tests for `build_format_from_order()` and `build_formats()`

dragosmg commented on r/tests/testthat/test-dplyr-funcs-datetime.R:
```diff
@@ -1965,3 +1971,339 @@ test_that("lubridate's fast_strptime", {
   )
 })
+
+test_that("parse_date_time with hours, minutes and seconds components", {
```
I've reorganised the giant block into multiple ones. Thanks :)

rok reviewed (approved):
This looks good to me (but I don't know the R codebase well)!
